File: android/app/src/main/kotlin/com/sovereign/communications/security/SecureDeletion.kt
package com.sovereign.communications.security

import java.io.File
import java.io.RandomAccessFile
import java.security.SecureRandom
import java.util.Arrays

/**
 * Secure deletion and memory wiping utilities for Android
 * 
 * IMPORTANT LIMITATIONS:
 * - SSDs use wear leveling which may prevent true data erasure
 * - File systems may keep journal copies
 * - JVM garbage collector may create copies that we cannot wipe
 * - These are BEST-EFFORT implementations for defense-in-depth
 * 
 * Primary defense should always be:
 * 1. Android Keystore for keys
 * 2. SQLCipher for database
 * 3. Encrypted storage
 */
object SecureDeletion {
    
    /**
     * Wipe a ByteArray by overwriting with zeros
     * 
     * Note: JVM garbage collector may have created copies we cannot wipe.
     * This is best-effort.
     */
    fun ByteArray.wipe() {
        Arrays.fill(this, 0.toByte())
    }
    
    /**
     * Wipe multiple ByteArrays
     */
    fun wipeMultiple(vararg arrays: ByteArray) {
        arrays.forEach { it.wipe() }
    }
    
    /**
     * Execute a block with automatic memory wiping
     * 
     * @param data Sensitive data to use
     * @param block Function to execute with the data
     * @return Result of the block
     */
    inline fun <T> withAutoWipe(data: ByteArray, block: (ByteArray) -> T): T {
        return try {
            block(data)
        } finally {
            data.wipe()
        }
    }
    
    /**
     * Securely delete a file by overwriting before deletion
     * 
     * LIMITATIONS:
     * - SSDs use wear leveling - data may persist on other blocks
     * - File system journaling may keep copies
     * - This is best-effort, not guaranteed secure erasure
     * 
     * @param file File to securely delete
     * @param passes Number of overwrite passes (default: 2)
     * @throws SecurityException if file cannot be wiped securely
     */
    fun secureDeleteFile(file: File, passes: Int = 2) {
        if (!file.exists()) {
            return
        }
        
        if (!file.isFile) {
            throw IllegalArgumentException("Can only securely delete files, not directories")
        }
        
        if (!file.canWrite()) {
            throw SecurityException("Cannot write to file for secure deletion: ${file.path}")
        }
        
        try {
            val fileSize = file.length()
            
            // Perform multiple overwrite passes
            repeat(passes) { pass ->
                RandomAccessFile(file, "rw").use { raf ->
                    raf.seek(0)
                    
                    // Buffer for writing
                    val bufferSize = minOf(4096, fileSize.toInt())
                    val buffer = ByteArray(bufferSize)
                    
                    var remaining = fileSize
                    
                    while (remaining > 0) {
                        val toWrite = minOf(bufferSize.toLong(), remaining).toInt()
                        
                        // Alternate between random and zeros
                        if (pass % 2 == 0) {
                            SecureRandom().nextBytes(buffer)
                        } else {
                            Arrays.fill(buffer, 0.toByte())
                        }
                        
                        raf.write(buffer, 0, toWrite)
                        remaining -= toWrite
                    }
                    
                    // Ensure data is written to disk
                    raf.fd.sync()
                }
            }
            
            // Final deletion
            if (!file.delete()) {
                throw SecurityException("Failed to delete file after wiping: ${file.path}")
            }
        } catch (e: Exception) {
            throw SecurityException("Secure deletion failed for ${file.path}", e)
        }
    }
    
    /**
     * Securely delete multiple files
     * 
     * @param files Files to delete
     * @param passes Number of overwrite passes per file
     * @return Map of files to their deletion success status
     */
    fun secureDeleteFiles(files: List<File>, passes: Int = 2): Map<File, Boolean> {
        return files.associateWith { file ->
            try {
                secureDeleteFile(file, passes)
                true
            } catch (e: Exception) {
                android.util.Log.e("SecureDeletion", "Failed to delete ${file.path}", e)
                false
            }
        }
    }
    
    /**
     * Securely delete a directory and all its contents
     * 
     * WARNING: This will recursively delete everything in the directory!
     * 
     * @param directory Directory to delete
     * @param passes Number of overwrite passes per file
     */
    fun secureDeleteDirectory(directory: File, passes: Int = 2) {
        if (!directory.exists()) {
            return
        }
        
        if (!directory.isDirectory) {
            throw IllegalArgumentException("Not a directory: ${directory.path}")
        }
        
        // Recursively delete contents
        directory.listFiles()?.forEach { file ->
            if (file.isDirectory) {
                secureDeleteDirectory(file, passes)
            } else {
                secureDeleteFile(file, passes)
            }
        }
        
        // Delete empty directory
        if (!directory.delete()) {
            throw SecurityException("Failed to delete directory: ${directory.path}")
        }
    }
    
    /**
     * Generate random bytes with automatic cleanup
     * 
     * @param size Size of random data
     * @param block Function to use the random data
     * @return Result of the block
     */
    inline fun <T> withRandomBytes(size: Int, block: (ByteArray) -> T): T {
        val data = ByteArray(size)
        SecureRandom().nextBytes(data)
        return try {
            block(data)
        } finally {
            data.wipe()
        }
    }
    
    /**
     * Timing-safe comparison with automatic wiping
     * 
     * @param a First array
     * @param b Second array
     * @param wipeAfter Whether to wipe arrays after comparison
     * @return true if arrays are equal
     */
    fun timingSafeCompareAndWipe(
        a: ByteArray,
        b: ByteArray,
        wipeAfter: Boolean = true
    ): Boolean {
        if (a.size != b.size) {
            if (wipeAfter) {
                wipeMultiple(a, b)
            }
            return false
        }
        
        var result = 0
        for (i in a.indices) {
            result = result or (a[i].toInt() xor b[i].toInt())
        }
        
        val isEqual = result == 0
        
        if (wipeAfter) {
            wipeMultiple(a, b)
        }
        
        return isEqual
    }
}

/**
 * Extension function for easier wiping syntax
 */
fun ByteArray.secureWipe() = SecureDeletion.run { this@secureWipe.wipe() }

/**
 * Extension function to securely delete a file
 */
fun File.secureDelete(passes: Int = 2) = SecureDeletion.secureDeleteFile(this, passes)

/**
 * Best practices and limitations documentation
 */
object SecureDeletionNotes {
    const val LIMITATIONS = """
        SECURE DELETION LIMITATIONS ON ANDROID:
        
        1. SSD/eMMC Wear Leveling:
           - Flash storage controllers remap blocks to distribute wear
           - Overwritten data may still exist on physical medium
           - True erasure requires full device encryption + factory reset
        
        2. File System Journaling:
           - ext4 and other filesystems keep transaction logs
           - Previous versions of files may exist in journal
           - Journal wiping requires root access
        
        3. Virtual Memory:
           - Data may be swapped to storage
           - We cannot control or wipe swap
           - Android encrypts swap on modern devices
        
        4. JVM Garbage Collection:
           - GC may create object copies we cannot track
           - Wiping is best-effort for JVM-managed memory
           - Sensitive data should use hardware-backed storage
        
        5. Copy-on-Write:
           - Some filesystems (like F2FS) use CoW
           - Modified blocks may leave old data intact
        
        BEST PRACTICES:
        
        1. PRIMARY DEFENSE: Use Android Keystore for all keys
        2. SECONDARY: Use SQLCipher for database encryption
        3. TERTIARY: Use this secure deletion as defense-in-depth
        4. DOCUMENT: Note limitations in security documentation
        5. EDUCATE: Users should enable device encryption + secure boot
        6. VERIFY: File deletion on device with appropriate security model
    """
    
    const val USAGE_EXAMPLE = """
        // Wipe sensitive data
        val password = "secret".toByteArray()
        try {
            authenticate(password)
        } finally {
            password.wipe()
        }
        
        // Use auto-wipe
        SecureDeletion.withAutoWipe(sensitiveData) { data ->
            process(data)
        }
        
        // Securely delete file
        val tempFile = File(cacheDir, "temp_secret.dat")
        try {
            tempFile.writeBytes(secretData)
            processFile(tempFile)
        } finally {
            tempFile.secureDelete()
        }
        
        // Generate random with auto-cleanup
        SecureDeletion.withRandomBytes(32) { nonce ->
            encrypt(message, nonce)
        }
    """
}

File: ios/SovereignCommunications/Security/SecureDeletion.swift
//
//  SecureDeletion.swift
//  Sovereign Communications
//
//  Secure deletion and memory wiping utilities for iOS
//
//  IMPORTANT LIMITATIONS:
//  - SSDs use wear leveling which may prevent true data erasure
//  - File systems may keep journal copies
//  - ARC may keep references we cannot control
//  - These are BEST-EFFORT implementations for defense-in-depth
//
//  Primary defense should always be:
//  1. iOS Keychain for keys
//  2. File protection (NSFileProtectionComplete)
//  3. Encrypted storage
//

import Foundation
import Security

/// Secure deletion and memory wiping utilities
class SecureDeletion {
    
    // MARK: - Memory Wiping
    
    /// Wipe a Data object by overwriting with zeros
    ///
    /// Note: ARC may keep references we cannot control.
    /// This is best-effort.
    ///
    /// - Parameter data: Data to wipe
    static func wipe(_ data: inout Data) {
        data.withUnsafeMutableBytes { ptr in
            guard let baseAddress = ptr.baseAddress else { return }
            memset(baseAddress, 0, ptr.count)
        }
    }
    
    /// Wipe multiple Data objects
    ///
    /// - Parameter dataObjects: Data objects to wipe
    static func wipeMultiple(_ dataObjects: inout [Data]) {
        for i in 0..<dataObjects.count {
            wipe(&dataObjects[i])
        }
    }
    
    /// Execute a block with automatic memory wiping
    ///
    /// - Parameters:
    ///   - data: Sensitive data to use
    ///   - block: Function to execute with the data
    /// - Returns: Result of the block
    static func withAutoWipe<T>(_ data: inout Data, block: (Data) throws -> T) rethrows -> T {
        defer {
            wipe(&data)
        }
        return try block(data)
    }
    
    // MARK: - Secure File Deletion
    
    /// Options for secure file deletion
    struct DeletionOptions {
        /// Number of overwrite passes (default: 2)
        let passes: Int
        
        /// Whether to verify deletion
        let verify: Bool
        
        init(passes: Int = 2, verify: Bool = false) {
            self.passes = passes
            self.verify = verify
        }
    }
    
    /// Securely delete a file by overwriting before deletion
    ///
    /// LIMITATIONS:
    /// - SSDs use wear leveling - data may persist on other blocks
    /// - File system journaling may keep copies
    /// - APFS uses copy-on-write which may leave old data
    /// - This is best-effort, not guaranteed secure erasure
    ///
    /// - Parameters:
    ///   - url: File URL to securely delete
    ///   - options: Deletion options
    /// - Throws: Error if deletion fails
    static func secureDelete(at url: URL, options: DeletionOptions = DeletionOptions()) throws {
        guard FileManager.default.fileExists(atPath: url.path) else {
            return
        }
        
        // Ensure it's a file, not a directory
        var isDirectory: ObjCBool = false
        guard FileManager.default.fileExists(atPath: url.path, isDirectory: &isDirectory),
              !isDirectory.boolValue else {
            throw SecureDeletionError.notAFile(url.path)
        }
        
        // Get file attributes
        let attributes = try FileManager.default.attributesOfItem(atPath: url.path)
        guard let fileSize = attributes[.size] as? UInt64 else {
            throw SecureDeletionError.cannotDetermineSize
        }
        
        // Open file for writing
        let fileHandle = try FileHandle(forWritingTo: url)
        defer {
            try? fileHandle.close()
        }
        
        // Perform overwrite passes
        for pass in 0..<options.passes {
            try fileHandle.seek(toOffset: 0)
            
            let bufferSize = min(4096, Int(fileSize))
            var buffer = Data(count: bufferSize)
            
            var remaining = fileSize
            
            while remaining > 0 {
                let toWrite = min(UInt64(bufferSize), remaining)
                
                // Alternate between random and zeros
                if pass % 2 == 0 {
                    // Random data
                    buffer.withUnsafeMutableBytes { ptr in
                        guard let baseAddress = ptr.baseAddress else { return }
                        _ = SecRandomCopyBytes(kSecRandomDefault, Int(toWrite), baseAddress)
                    }
                } else {
                    // Zero data
                    buffer.withUnsafeMutableBytes { ptr in
                        guard let baseAddress = ptr.baseAddress else { return }
                        memset(baseAddress, 0, Int(toWrite))
                    }
                }
                
                try fileHandle.write(contentsOf: buffer.prefix(Int(toWrite)))
                remaining -= toWrite
            }
            
            // Ensure data is written to disk
            try fileHandle.synchronize()
        }
        
        // Close file before deletion
        try fileHandle.close()
        
        // Final deletion
        try FileManager.default.removeItem(at: url)
        
        // Verify deletion if requested
        if options.verify && FileManager.default.fileExists(atPath: url.path) {
            throw SecureDeletionError.deletionFailed(url.path)
        }
    }
    
    /// Securely delete multiple files
    ///
    /// - Parameters:
    ///   - urls: File URLs to delete
    ///   - options: Deletion options
    /// - Returns: Dictionary of URLs to their deletion success status
    static func secureDeleteMultiple(
        _ urls: [URL],
        options: DeletionOptions = DeletionOptions()
    ) -> [URL: Result<Void, Error>] {
        var results: [URL: Result<Void, Error>] = [:]
        
        for url in urls {
            do {
                try secureDelete(at: url, options: options)
                results[url] = .success(())
            } catch {
                results[url] = .failure(error)
            }
        }
        
        return results
    }
    
    /// Securely delete a directory and all its contents
    ///
    /// WARNING: This will recursively delete everything in the directory!
    ///
    /// - Parameters:
    ///   - url: Directory URL to delete
    ///   - options: Deletion options for files
    /// - Throws: Error if deletion fails
    static func secureDeleteDirectory(
        at url: URL,
        options: DeletionOptions = DeletionOptions()
    ) throws {
        guard FileManager.default.fileExists(atPath: url.path) else {
            return
        }
        
        var isDirectory: ObjCBool = false
        guard FileManager.default.fileExists(atPath: url.path, isDirectory: &isDirectory),
              isDirectory.boolValue else {
            throw SecureDeletionError.notADirectory(url.path)
        }
        
        // Get directory contents
        let contents = try FileManager.default.contentsOfDirectory(
            at: url,
            includingPropertiesForKeys: [.isDirectoryKey],
            options: []
        )
        
        // Recursively delete contents
        for itemURL in contents {
            let resourceValues = try itemURL.resourceValues(forKeys: [.isDirectoryKey])
            
            if resourceValues.isDirectory == true {
                try secureDeleteDirectory(at: itemURL, options: options)
            } else {
                try secureDelete(at: itemURL, options: options)
            }
        }
        
        // Delete empty directory
        try FileManager.default.removeItem(at: url)
    }
    
    // MARK: - Utility Functions
    
    /// Generate random data with automatic cleanup
    ///
    /// - Parameters:
    ///   - size: Size of random data in bytes
    ///   - block: Function to use the random data
    /// - Returns: Result of the block
    static func withRandomData<T>(size: Int, block: (Data) throws -> T) rethrows -> T {
        var data = Data(count: size)
        defer {
            wipe(&data)
        }
        
        data.withUnsafeMutableBytes { ptr in
            guard let baseAddress = ptr.baseAddress else { return }
            _ = SecRandomCopyBytes(kSecRandomDefault, size, baseAddress)
        }
        
        return try block(data)
    }
    
    /// Timing-safe comparison with automatic wiping
    ///
    /// - Parameters:
    ///   - a: First data
    ///   - b: Second data
    ///   - wipeAfter: Whether to wipe data after comparison
    /// - Returns: true if data objects are equal
    static func timingSafeCompareAndWipe(
        _ a: inout Data,
        _ b: inout Data,
        wipeAfter: Bool = true
    ) -> Bool {
        defer {
            if wipeAfter {
                wipe(&a)
                wipe(&b)
            }
        }
        
        guard a.count == b.count else {
            return false
        }
        
        var result: UInt8 = 0
        
        a.withUnsafeBytes { aPtr in
            b.withUnsafeBytes { bPtr in
                guard let aBase = aPtr.baseAddress?.assumingMemoryBound(to: UInt8.self),
                      let bBase = bPtr.baseAddress?.assumingMemoryBound(to: UInt8.self) else {
                    return
                }
                
                for i in 0..<a.count {
                    result |= aBase[i] ^ bBase[i]
                }
            }
        }
        
        return result == 0
    }
}

// MARK: - Data Extension

extension Data {
    /// Wipe this Data object by overwriting with zeros
    mutating func secureWipe() {
        SecureDeletion.wipe(&self)
    }
}

// MARK: - URL Extension

extension URL {
    /// Securely delete the file at this URL
    ///
    /// - Parameter options: Deletion options
    /// - Throws: Error if deletion fails
    func secureDelete(options: SecureDeletion.DeletionOptions = SecureDeletion.DeletionOptions()) throws {
        try SecureDeletion.secureDelete(at: self, options: options)
    }
}

// MARK: - Errors

enum SecureDeletionError: Error, LocalizedError {
    case notAFile(String)
    case notADirectory(String)
    case cannotDetermineSize
    case deletionFailed(String)
    
    var errorDescription: String? {
        switch self {
        case .notAFile(let path):
            return "Not a file: \(path)"
        case .notADirectory(let path):
            return "Not a directory: \(path)"
        case .cannotDetermineSize:
            return "Cannot determine file size"
        case .deletionFailed(let path):
            return "Deletion verification failed for: \(path)"
        }
    }
}

// MARK: - Best Practices Documentation

enum SecureDeletionNotes {
    static let limitations = """
        SECURE DELETION LIMITATIONS ON iOS:
        
        1. SSD/Flash Wear Leveling:
           - Flash storage controllers remap blocks to distribute wear
           - Overwritten data may still exist on physical medium
           - True erasure requires full device encryption + erase all content
        
        2. APFS Copy-on-Write:
           - APFS filesystem uses CoW for many operations
           - Modified blocks may leave old data intact
           - Snapshots may preserve old versions
        
        3. File System Journaling:
           - HFS+ and APFS keep transaction logs
           - Previous versions of files may exist in journal
        
        4. Virtual Memory:
           - Data may be swapped to storage
           - We cannot control or wipe swap
           - iOS encrypts swap on all modern devices
        
        5. ARC Memory Management:
           - ARC may create object copies we cannot track
           - Wiping is best-effort for ARC-managed memory
           - Sensitive data should use Keychain
        
        BEST PRACTICES:
        
        1. PRIMARY DEFENSE: Use iOS Keychain for all keys
        2. SECONDARY: Use file protection (NSFileProtectionComplete)
        3. TERTIARY: Use this secure deletion as defense-in-depth
        4. DOCUMENT: Note limitations in security documentation
        5. EDUCATE: Users should enable device passcode + File Data Protection
        6. EXCLUDE: Exclude sensitive files from backups
        """
    
    static let usageExample = """
        // Wipe sensitive data
        var password = "secret".data(using: .utf8)!
        defer {
            password.secureWipe()
        }
        try authenticate(password)
        
        // Use auto-wipe
        try SecureDeletion.withAutoWipe(&sensitiveData) { data in
            try process(data)
        }
        
        // Securely delete file
        let tempFile = FileManager.default.temporaryDirectory
            .appendingPathComponent("temp_secret.dat")
        
        try secretData.write(to: tempFile)
        defer {
            try? tempFile.secureDelete()
        }
        try processFile(tempFile)
        
        // Generate random with auto-cleanup
        let result = try SecureDeletion.withRandomData(size: 32) { nonce in
            try encrypt(message, nonce: nonce)
        }
        """
}
